src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEntry.Unix.cs

```diff
@@ -34,52 +33,44 @@ namespace System.IO.Enumeration
             entry._fullPath = ReadOnlySpan<char>.Empty;
             entry._fileName = ReadOnlySpan<char>.Empty;
 
-            // IMPORTANT: Attribute logic must match the logic in FileStatus
+            entry._status = default;
+            entry._status.InvalidateCaches();
+
+            bool isDirectory = directoryEntry.InodeType == Interop.Sys.NodeType.DT_DIR;
+            bool isSymlink = directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK;
 
-            bool isDirectory = false;
-            bool isSymlink = false;
-            if (directoryEntry.InodeType == Interop.Sys.NodeType.DT_DIR)
-            {
-                // We know it's a directory.
-                isDirectory = true;
-            }
             // Some operating systems don't have the inode type in the dirent structure,
             // so we use DT_UNKNOWN as a sentinel value. As such, check if the dirent is a
             // directory.
-            else if ((directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK
-                || directoryEntry.InodeType == Interop.Sys.NodeType.DT_UNKNOWN)
-                && Interop.Sys.Stat(entry.FullPath, out Interop.Sys.FileStatus targetStatus) >= 0)
-            {
-                // Symlink or unknown: Stat to it to see if we can resolve it to a directory.
-                isDirectory = (targetStatus.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFDIR;
-            }
-            // Same idea as the directory check, just repeated for (and tweaked due to the
-            // nature of) symlinks.
-            if (directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK)
+            if (!isDirectory &&
+                (directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK ||
+                 directoryEntry.InodeType == Interop.Sys.NodeType.DT_UNKNOWN))
             {
-                isSymlink = true;
+                // DT_LNK: stat to it to see if we can resolve it to a directory
+                entry._status.TryRefreshSymbolicLinkCache(entry.FullPath);
+                isDirectory = entry._status.HasDirectoryFlag;
             }
-            else if ((directoryEntry.InodeType == Interop.Sys.NodeType.DT_UNKNOWN)
-                && (Interop.Sys.LStat(entry.FullPath, out Interop.Sys.FileStatus linkTargetStatus) >= 0))
+
+            if (!isSymlink &&
+                directoryEntry.InodeType == Interop.Sys.NodeType.DT_UNKNOWN)
```

Comment:
```suggestion
            if (!isSymlink && isUnknown)
```

As mentioned above, they are mutually exclusive, so this could potentially be simplified:

```suggestion
            if (isUnknown)
```